### PR TITLE
Add Connect protocol implementation

### PR DESCRIPTION
**Pull request by akshayjshah**
Once all the prefactoring in #212 lands, it doesn't take much to add an
implementation of the Connect protocol: `protocol_connect.go` clocks in at just
809 lines of Go. Clients default to using the Connect protocol, and handlers
automatically support it.

### CODE REVIEW

File: protocol_connect.go [hunk 1]
```diff
@@ -0,0 +1,963 @@
...
+	"time"
+
+	errorv1 "github.com/bufbuild/connect-go/internal/gen/go/connect/error/v1"
+)
+
+const (
+	connectUnaryHeaderCompression           = "Content-Encoding"
+	connectUnaryHeaderAcceptCompression     = "Accept-Encoding"
+	connectUnaryTrailerPrefix               = "Trailer-"
+	connectStreamingHeaderCompression       = "Connect-Encoding"
```
Comment (buildbreaker) [inline protocol_connect.go:37]: Maybe we can converge content-encoding for both unary and streaming since they are using the same key value?

**Review by buildbreaker — APPROVED** [review]
Nice!

I like that the protocol is very concise code wise but do we have any plans on breaking out some the `protocol_connect.go` file down a bit? It's nice to have as a reference but it was difficult for me to immediately see the client specific code versus the handler specific code

**Comment by buildbreaker** [comment]
Thanks for the explanation!

**Review by bufdev — APPROVED** [review]
All looks sane to me - let's go for it.

The only high-level comment is I wish we had a few more integration tests that were protocol-independent - maybe I just need to look more closely, but I think I was expecting more to be here. I know connect-crosstest has a lot, but some stuff that lives here for on-lock verification would be good.

**Comment by akshayjshah** [comment]
Most of the test coverage for this repo is protocol-independent - see TestServer in connect_ext_test.go. It doesn't show up in this diff though - none of it needed to change. There's always more cases we can add though :)

We shouldn't be relying on crosstest for anything that can be tested without a gRPC-Go dependency. Any bugs we've found with crosstest have had pure-connect regression tests added here as part of the fix.

File: protocol_connect.go [hunk 2]
```diff
@@ -0,0 +1,963 @@
...
+	request     *http.Request
+}
+
+func (r *connectStreamingHandlerReceiver) Receive(message any) error {
+	if err := r.unmarshaler.Unmarshal(message); err != nil {
+		// Clients may not send end-of-stream metadata, so we don't need to handle
+		// errSpecialEnvelope.
+		return err
+	}
+	return nil // no typed nils
```
[the remark this answers is not in the document]
Comment (akshayjshah) [reply protocol_connect.go:480]: Sure! Would you prefer a longer explanation everywhere or a one-time explanation somewhere?

[the remark this answers is not in the document]
Comment (bufdev) [reply protocol_connect.go:480]: Whatever you think, just at least something to warn off future developers from trying to simplify this

**Pull request merged**